File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, a: List[List[str]]) -> int:
        n = len(a)
        m = len(a[0])
        
        visited = [[0 for j in range(m)] for i in range(n)]
        
        dr = [0,0,-1,1]
        dc = [-1,1,0,0]
        
        dq = deque()
        ans = 0
        print(visited)
        print(a)
        for i in range(n):
            for j in range(m):
                if a[i][j]=='0': continue
                if visited[i][j]: continue
                dq.append((i,j))                
                while dq:
                    r,c = dq.pop()
                    for k in range(4):
                        nr = r + dr[k]
                        nc = c + dc[k]                                
                        if nr < 0 or nr >= n: continue
                        if nc < 0 or nc >= m: continue
                        if visited[nr][nc]: continue
                        if a[nr][nc] == '0': continue
                        visited[nr][nc] = 1
                        dq.append((nr,nc))
                ans += 1
        return ans
                

File: 0200-number-of-islands/0200-number-of-islands.cpp
using vi = vector<int>;
using vvi = vector<vi>;
vi dr = {0,0,-1,1};
vi dc = {-1,1,0,0};

class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        vvi visited(n, vi(m));

        int ans = 0;
        for (int i=0; i<n; i++) {
            for (int j=0; j<m; j++) {
                if (visited[i][j]==1) continue;
                if (grid[i][j] == '0') continue;

                ans++;
                queue<pair<int,int>> que;
                que.push({i,j});
                visited[i][j] = 1;
                while (!que.empty()) {
                    auto [r,c] = que.front();
                    que.pop();
                    for (int k=0; k<4; k++) {
                        int nr = r + dr[k];
                        int nc = c + dc[k];
                        if (nr < 0 || nr >= n) continue;
                        if (nc < 0 || nc >= m) continue;
                        if (visited[nr][nc]==1) continue;
                        if (grid[nr][nc] == '0') continue;
                        visited[nr][nc] = 1;
                        que.push({nr,nc});
                    }
                }
            }
        }
        return ans;
    }
};